app/src/main/java/org/dhis2/usescases/searchTrackEntity/SearchTEIViewModel.kt

```diff
@@ -0,0 +1,268 @@
+            !_allowCreateWithoutSearch &&
+            !searching
+        _screenState.value = when {
+            shouldOpenSearch ->
+                SearchForm(
+                    previousSate = _screenState.value?.screenState ?: SearchScreenState.LIST,
+                    queryHasData = queryData.isNotEmpty(),
+                    minAttributesToSearch = _selectedProgram.value?.minAttributesRequiredToSearch()
+                        ?: 1
+                )
+            else ->
+                SearchList(
+                    previousSate = _screenState.value?.screenState ?: SearchScreenState.NONE,
+                    displayFrontPageList = _selectedProgram.value?.displayFrontPageList() ?: false,
+                    canCreateWithoutSearch = _allowCreateWithoutSearch,
+                    isSearching = searching
+                )
+        }
+    }
+
+    fun setMapScreen() {
+        _screenState.value = SearchMap(
+            previousSate = _screenState.value?.screenState ?: SearchScreenState.NONE,
+            canCreateWithoutSearch = _allowCreateWithoutSearch
+        )
+    }
+
+    fun setAnalyticsScreen() {
+        _screenState.value = SearchAnalytics(
+            previousSate = _screenState.value?.screenState ?: SearchScreenState.NONE
+        )
+    }
+
+    fun setSearchScreen() {
+        _screenState.value = SearchForm(
+            previousSate = _screenState.value?.screenState ?: SearchScreenState.NONE,
+            queryHasData = queryData.isNotEmpty(),
+            minAttributesToSearch = _selectedProgram.value?.minAttributesRequiredToSearch() ?: 0
+        )
+    }
+
+    fun setPreviousScreen() {
+        when (_screenState.value?.previousSate) {
+            SearchScreenState.LIST -> setListScreen()
+            SearchScreenState.MAP -> setMapScreen()
+            SearchScreenState.SEARCHING -> setSearchScreen()
+            SearchScreenState.ANALYTICS -> setAnalyticsScreen()
+            else -> {
+            }
+        }
+    }
+
+    fun refreshData() {
+        onSearchClick()
+    }
+
+    fun updateQueryData(rowAction: RowAction) {
+        if (rowAction.value != null) {
+            queryData[rowAction.id] = rowAction.value!!
+        } else {
+            queryData.remove(rowAction.id)
+        }
+        updateSearch()
+    }
+
+    fun clearQueryData() {
+        queryData.clear()
+        updateSearch()
+    }
+
+    fun restoreQueryData(queryDataBackUp: HashMap<String, String>) {
+        queryData.clear()
+        queryData.putAll(queryDataBackUp)
+        updateSearch()
+    }
+
+    private fun updateSearch() {
+        if (_screenState.value is SearchForm) {
+            _screenState.value =
+                (_screenState.value as SearchForm).copy(queryHasData = queryData.isNotEmpty())
+        }
+    }
+
+    fun fetchListResults(): LiveData<PagedList<SearchTeiModel>>? {
+        return when {
+            searching -> {
+                searchRepository.searchTrackedEntities(
```

Comment:
same here, are these calls to D2 blocking the main thread? should these be async?